proposed/0000-UDP-hole-punching.md

```diff
@@ -0,0 +1,192 @@
+server to tell each peer's their external UDP endpoints. Since our intent
+is a fully decentralised system, this text also describes how nodes to which
+we're already connected can be used as a replacement for such server.
+
+In the following text the reader is expected to know basic NAT types and how they work.
+The section 'Methods of translation' on [wikipedia](https://en.wikipedia.org/wiki/Network_address_translation)
+gives a nice overview.
+
+# Detailed design
+
+The UDP hole punching process described here consists of two steps:
+
+1. Finding an external mapping of a UDP socket
+2. Do the actual hole punching.
+
+## Finding an external mapping of a UDP socket
+
+Suppose, we have two nodes `A` and `B`, where each one is possibly behind a NAT and
+they want to connect to each other.  Routing needs to send `A` an endpoint of `B` and
+vice versa. But `A` and `B` don’t know their public
+endpoints a priori (they may know their IP addresses from previous connection establishments,
+but not their ports). Each node therefore needs to create a UDP socket and use it to ask some
+other node it is connected to what the public endpoints of the UDP socket is. We’ll refer
+to such helping nodes `C(A)` and `C(B)` (it needs not be the same node for `A` and `B`).
+We'll call `U(C(X))` the UDP port on which X can contact `C(X)` in order to find its public
+endpoint. `X` finds out about `U(C(X))` through the initial handhsake, that is, when
+`X` connects to `C(X)`.
+
+In the following text we’ll describe steps taken at node `A`, but the same steps need to
+be taken at node `B`.
+
+For `A` to find out its external address `A` first needs to create a UDP socket `S(A)`
+with an arbitrary local port.  `A` will then use `S(A)` to periodically send
+datagrams to `U(C(A))` containing a MAGIC number and the ID of the request. Each time
+`C(A)` receives such datagram, it sends back a datagram containing public endpoint
+of `S(A)` together with the request ID.
+
+The above protocol shall be initiated from upper layers by calling
+
+    Service::get_mapped_udp_socket(callback: FnBox<(io::Result<MappedUdpSocket>)>)
+
+And the result of such call shall be an invocation of the callback.
+
+where `MappedUdpSocket` is defined as
+
+    struct MappedUdpSocket {
+        udp_socket : UdpSocket,
+        public_address : SocketAddr, // of node A
+    }
+
+Once upper layers receive such event, they can send/route `MappedUdpSocket::public_address`
+to node `B`. Once `B` does the same, and upper layers receive `B`’s public endpoint, upper
+layers are ready for the actual hole punching.
+
+## Hole punching
+
+The act of hole punching shall be initiated by a function with the following signature:
+
+    Service::udp_punch_hole(udp_socket : UdpSocket,
+                            secret: Option<[u8; 4]>,
+                            peer_addr : mut SocketAddr /* of node B */,
+                            callback: FnBox<(UdpSocket, io::Result<SocketAddr>)>)
```

Comment:
`Service::udp_punch_hole` doesn't know enough info to initiate a connection. Currently, only one UDP-based protocol is supported, then it's not a problem. But given this function will be used by upper layers, we should do a little more to avoid breaking API as soon as we implement rUDP or whatever other custom protocol builtin on top of UDP.

What about adding another param like `Protocol`? Given not all `Protocol`s are UDP-based, we could have a new enum, `UdpProtocol`.